schemaregistry/internal/rest_service.go

```diff
@@ -155,14 +157,73 @@ func NewRestService(conf *ClientConfig) (*RestService, error) {
 		}
 	}
 
+	var authenticationHeaderProvider AuthenticationHeaderProvider = nil
+
+	if conf.BearerAuthCredentialsSource == "OAUTHBEARER" {
+		if conf.AuthenticationHeaderProvider != nil {
+			return nil, fmt.Errorf("cannot have bearer.auth.credentials.source oauthbearer " +
+				"with custom authentication header provider")
+		}
+
+		if conf.BearerAuthIssuerEndpointURL == "" {
+			return nil, fmt.Errorf("bearer.auth.issuer.endpoint.url must be specified when bearer.auth.credentials.source is" +
+				" specified with OAUTHBEARER")
+		}
+
+		if conf.BearerAuthClientID == "" {
+			return nil, fmt.Errorf("bearer.auth.client.id must be specified when bearer.auth.credentials.source is" +
+				" specified with OAUTHBEARER")
+		}
+
+		if conf.BearerAuthClientSecret == "" {
+			return nil, fmt.Errorf("bearer.auth.client.secret must be specified when bearer.auth.credentials.source is" +
+				" specified with OAUTHBEARER")
+		}
+
+		if len(conf.BearerAuthScopes) == 0 {
+			return nil, fmt.Errorf("bearer.auth.scopes must be specified when bearer.auth.credentials.source is" +
+				" specified with OAUTHBEARER")
+		}
+		//TODO: For major version release, lsrc + identity pool id should be required for static token as well
+		if conf.BearerAuthLogicalCluster == "" {
+			return nil, fmt.Errorf("bearer.auth.logical.cluster must be specified when bearer.auth.credentials.source is" +
+				" specified with OAUTHBEARER")
+		}
+
+		if conf.BearerAuthIdentityPoolID == "" {
+			return nil, fmt.Errorf("bearer.auth.identity.pool.id must be specified when bearer.auth.credentials.source is" +
+				" specified with OAUTHBEARER")
+		}
+
+		tokenFetcher := &clientcredentials.Config{
+			ClientID:     conf.BearerAuthClientID,
+			ClientSecret: conf.BearerAuthClientSecret,
+			TokenURL:     conf.BearerAuthIssuerEndpointURL,
+			Scopes:       conf.BearerAuthScopes,
+		}
+		authenticationHeaderProvider = NewBearerTokenAuthenticationHeaderProvider(
+			tokenFetcher,
+			conf.MaxRetries,
+			conf.RetriesWaitMs,
+			conf.RetriesMaxWaitMs,
+		)
+	} else if conf.BearerAuthCredentialsSource == "CUSTOM" {
+		if conf.AuthenticationHeaderProvider == nil {
+			return nil, fmt.Errorf("cannot have bearer.auth.credentials.source custom " +
+				"with no custom authentication header provider")
+		}
+		authenticationHeaderProvider = conf.AuthenticationHeaderProvider
+	}
+
 	return &RestService{
-		urls:             urls,
-		headers:          headers,
-		maxRetries:       conf.MaxRetries,
-		retriesWaitMs:    conf.RetriesWaitMs,
-		retriesMaxWaitMs: conf.RetriesMaxWaitMs,
-		ceilingRetries:   int(math.Log2(float64(conf.RetriesMaxWaitMs) / float64(conf.RetriesWaitMs))),
-		Client:           conf.HTTPClient,
+		urls:                         urls,
+		headers:                      headers,
```

Comment:
It's a bit messy that the auth headers are already set in the case of basic auth, whereas for OAuth we use the `authenticationHeaderProvider`.  It would be cleaner if the only headers that were set were possibly the `target-sr-cluster` and the `Confluent-Identity-Pool-Id`, and then the `authenticationHeaderProvider` would be one that could set the auth header for both basic auth and OAuth.  WDYT?